hask/02.hs

```diff
@@ -20,25 +20,20 @@ main = do
     print $ solve part1 contents
     print $ solve part2 contents
 
-solve f = sum . map (line f) . zip [1..] . lines
+solve f = sum . map f . zip [1..] . lines
 
-line :: (Eq a, Eq b, Num a) => ((a,b) -> Maybe a) -> (a,b) -> a
-line f s | res == Nothing = 0
-         | otherwise = fromJust res
-    where res = f s
+validColor c x =
+    last s /= c ||
+    read (head s) <= getMaxColor c
+  where s = words x
```

Comment:
Just the usual trick of replacing statements with a boolean expression for functions that return booleans. If it got too complicated you could leave the structure.